src/Microsoft.Diagnostics.Runtime/src/Common/ClrArray.cs

```diff
@@ -127,6 +117,139 @@ public int GetLength(int dimension)
+                throw new ArgumentException("Array was not a one-dimensional array.");
+
+            int valueOffset = index;
+            int dataByteOffset = 2 * IntPtr.Size;
+
+            if (IsMultiDimensional)
+            {
+                valueOffset -= GetMultiDimensionalBound(1);
+                if ((uint)valueOffset >= GetMultiDimensionalBound(0))
+                    throw new ArgumentOutOfRangeException(nameof(index));
+
+                dataByteOffset += 2 * sizeof(int);
+            }
+            else
+            {
+                if ((uint)valueOffset >= Length)
+                    throw new ArgumentOutOfRangeException(nameof(index));
+            }
+
+            int valueByteOffset = dataByteOffset + valueOffset * Type.ComponentSize;
+            return Type.ClrObjectHelpers.DataReader.Read<T>(Address + (ulong)valueByteOffset);
+        }
+
+        public T GetValue<T>(params int[] indices) where T : unmanaged
+        {
+            if (indices is null)
+                throw new ArgumentNullException(nameof(indices));
+
+            int rank = Rank;
+            if (rank != indices.Length)
+                throw new ArgumentException("Indices length does not match the array rank.");
+
+            int valueOffset = 0;
+            int dataByteOffset = 2 * IntPtr.Size;
+
+            if (rank == 1)
+            {
+                if (IsMultiDimensional)
+                {
+                    valueOffset = indices[0] - GetMultiDimensionalBound(1);
+                    if ((uint)valueOffset >= GetMultiDimensionalBound(0))
+                        throw new ArgumentOutOfRangeException(nameof(indices));
+
+                    dataByteOffset += 2 * sizeof(int);
+                }
+                else
+                {
+                    valueOffset = indices[0];
+                    if ((uint)valueOffset >= Length)
+                        throw new ArgumentOutOfRangeException(nameof(indices));
+                }
+            }
+            else
+            {
+                for (int dimension = 0; dimension < rank; dimension++)
+                {
+                    int currentValueOffset = indices[dimension] - GetMultiDimensionalBound(rank + dimension);
+                    if ((uint)currentValueOffset >= GetMultiDimensionalBound(dimension))
+                        throw new ArgumentOutOfRangeException(nameof(indices));
+
+                    valueOffset *= GetMultiDimensionalBound(dimension);
+                    valueOffset += currentValueOffset;
+                }
+
+                dataByteOffset += 2 * sizeof(int) * rank;
+            }
+
+            int valueByteOffset = dataByteOffset + valueOffset * Type.ComponentSize;
+            return Type.ClrObjectHelpers.DataReader.Read<T>(Address + (ulong)valueByteOffset);
+        }
+
+        public ClrObject GetObjectValue(int index)
+        {
+            ulong address = GetValue<UIntPtr>(index).ToUInt64();
+            if (!Type.IsObjectReference)
```

Comment:
nit:  Maybe this should go above the read?

Reply:
Same for GetObjectValue.  It seems like we should throw an InvalidOperationException if the type isn't an object reference unconditionally.  It's possible that GetValue might throw a different exception depending on the underlying IMemoryReader implementation, so we should throw the InvalidOperationException as a priority.